src/main/java/me/zeroeightsix/kami/module/modules/player/ViewLock.kt

```diff
@@ -0,0 +1,117 @@
+
+import me.zeroeightsix.kami.event.events.SafeTickEvent
+import me.zeroeightsix.kami.module.Module
+import me.zeroeightsix.kami.setting.Setting
+import me.zeroeightsix.kami.setting.Settings
+import me.zeroeightsix.kami.util.event.listener
+import java.util.*
+import kotlin.math.abs
+import kotlin.math.round
+import kotlin.math.sign
+
+@Module.Info(
+        name = "ViewLock",
+        category = Module.Category.PLAYER,
+        description = "Locks your camera view"
+)
+object ViewLock : Module() {
+
+    val yaw = register(Settings.b("Yaw", true))
+    val pitch = register(Settings.b("Pitch", true))
+    val autoYaw = register(Settings.booleanBuilder("AutoYaw").withValue(true).withVisibility { yaw.value })
+    val autoPitch = register(Settings.booleanBuilder("AutoPitch").withValue(true).withVisibility { pitch.value })
+    private val threshold = register(Settings.integerBuilder("ChangeThresh").withValue(5).withRange(1, 50).withStep(1).withVisibility { autoPitch.value || autoYaw.value })
+    val disableMouseYaw = register(Settings.booleanBuilder("DisableMouseYaw").withValue(true).withVisibility { yaw.value })
+    val disableMousePitch = register(Settings.booleanBuilder("DisableMousePitch").withValue(true).withVisibility { pitch.value })
+    private val specifcYaw = register(Settings.floatBuilder("SpecificYaw").withValue(180.0f).withRange(-180.0f, 180.0f).withStep(1.0f).withVisibility { !autoYaw.value })
+    private val specificPitch = register(Settings.floatBuilder("SpecificPitch").withValue(0.0f).withRange(-90.0f, 90.0f).withStep(1.0f).withVisibility { !autoPitch.value })
+    private val yawSlice = register(Settings.integerBuilder("YawSlice").withValue(8).withRange(2, 32).withStep(1).withVisibility { autoYaw.value })
+    private val pitchSlice = register(Settings.integerBuilder("PitchSlice").withValue(5).withRange(2, 32).withStep(1).withVisibility { autoPitch.value })
+
+    private var yawSanp = 0
+    private var pitchSnap = 0
+    private val deltaXQueue = LinkedList<Pair<Int, Long>>()
+    private val deltaYQueue = LinkedList<Pair<Int, Long>>()
+    private var pitchSliceAngle = 1.0f
+    private var yawSliceAngle = 1.0f
+
+    override fun onEnable() {
+        yawSliceAngle = 360.0f / yawSlice.value
+        pitchSliceAngle = 180.0f / (pitchSlice.value - 1)
+        if ((autoYaw.value || autoPitch.value) && mc.player != null)
+            snapToNext()
+    }
+
+    init {
+        listener<SafeTickEvent> {
+            if (autoYaw.value || autoPitch.value) {
+                snapToSlice()
+            }
+            if (yaw.value && !autoYaw.value) mc.player.rotationYaw = specifcYaw.value
+            if (pitch.value && !autoPitch.value) mc.player.rotationPitch = specificPitch.value
+        }
+    }
+
+    private fun snapToNext() {
+        yawSanp = round(mc.player.rotationYaw / yawSliceAngle).toInt()
+        pitchSnap = round((mc.player.rotationPitch + 90) / pitchSliceAngle).toInt()
```

Comment:
This seems awfully imprecise, is this jarring to the user?

Reply:
this just calculates to which direction the view should snap. the rounding is needed here to find the next step to snap to